### DISCUSSION TITLE
How to SetValue with array properly?

### DISCUSSION THREAD

**Discussion by sea0920** [sebastienros/jint#1297] [Q&A]
I thought I had to set ArrayInstance when I use SetValue.
But it seems like I have to put object[]?
Otherwise, I cannot use push() method.

```
class Program
    {
        static void Main(string[] args)
        {
            var engine = new Jint.Engine(cfg => cfg.AllowClr());

            var dict = new Dictionary<uint, Jint.Runtime.Descriptors.PropertyDescriptor>();
            dict.Add(0, new Jint.Runtime.Descriptors.PropertyDescriptor(1, true, true, true));
            dict.Add(1, new Jint.Runtime.Descriptors.PropertyDescriptor(2, true, true, true));
            var myArr = new ArrayInstance(engine, dict);
            engine.SetValue("myArr", myArr); // myArr.push(3) works when I change myArr to myArr.ToObject()
            engine.Execute("myArr.push(3); System.Console.WriteLine(myArr[2]);");
        }
    }
```

**Comment by lahma** [comment]
So what are you trying to achieve? Do you want to fill the array on c# side or what's this use case. It should be enough just to create and array of objects like `new object[] { 123, "hello world" }` and give that to SetValue. I hope you are using v3 beta, so we are talking about the same API.

**Reply by sea0920** [reply]
I am using v3 beta. I am not clear why we need a class like ArrayInstance. Is it something used internally and Jint users don't need to know? Does the array of objects `new object[] { 123, "hello world" }` turn into ArrayInstance when we SetValue?

**Reply by lahma** [reply]
`ArrayInstance` generally is used by Jint during runtime, but it's also available to C# when needed. The example array should turn into `ArrayInstance` when used with `SetValue`, all elements in array also go through the `JsValue` conversion during the process. After that the array inside the engine "native-like".
